File: halstead_metrics.php
<?php
function calculateHalsteadMetrics($code) {
    $operators = [];
    $operands = [];
    $operatorCount = 0;
    $operandCount = 0;

    // Tokenize the code
    $tokens = token_get_all($code);

    foreach ($tokens as $token) {
        if (is_array($token)) {
            [$tokenType, $tokenValue] = $token;

            // Operators (using numeric token values)
            if ($tokenType === 277 || // T_PLUS_EQUAL
                $tokenType === 276 || // T_MINUS_EQUAL
                $tokenType === 275 || // T_MUL_EQUAL
                $tokenType === 274 || // T_DIV_EQUAL
                $tokenType === 272 || // T_MOD_EQUAL
                $tokenType === 271 || // T_EQUAL
                $tokenType === 260 || // T_DOUBLE_ARROW
                $tokenType === 279 || // T_BOOLEAN_AND
                $tokenType === 278) { // T_BOOLEAN_OR
                $operators[$tokenValue] = ($operators[$tokenValue] ?? 0) + 1;
                $operatorCount++;
            }
            // Operands (variables, constants, etc.)
            elseif ($tokenType === 309 || // T_VARIABLE
                    $tokenType === 307) { // T_STRING
                $operands[$tokenValue] = ($operands[$tokenValue] ?? 0) + 1;
                $operandCount++;
            }
        }
    }

    // Halstead's metrics
    $n1 = count($operators); // Number of distinct operators
    $n2 = count($operands);  // Number of distinct operands
    $N1 = $operatorCount;    // Total number of operators
    $N2 = $operandCount;     // Total number of operands

    // Program vocabulary
    $n = $n1 + $n2;
    // Program length
    $N = $N1 + $N2;
    // Volume
    $V = $N * log($n, 2);
    // Difficulty
    $D = ($n1 / 2) * ($N2 / $n2);
    // Effort
    $E = $D * $V;

    return [
        "n1" => $n1,
        "n2" => $n2,
        "N1" => $N1,
        "N2" => $N2,
        "n" => $n,
        "N" => $N,
        "V" => $V,
        "D" => $D,
        "E" => $E
    ];
}

$code = <<<'CODE'
<?php
function add($a, $b) {
    return $a + $b;
}

function subtract($a, $b) {
    return $a - $b;
}
CODE;

$metrics = calculateHalsteadMetrics($code);
echo "Halstead's Metrics:\n";
foreach ($metrics as $key => $value) {
    echo "$key: $value\n";
}
?>

File: halstead_metrics.py
import os
import re

def extract_operators_and_operands(code):
    operators = re.findall(r'[+\-*/%=<>!&|^~?:]+|\b(function|return|if|else|for|while|switch|case|break|continue|new|var|let|const|class|extends|import|export|try|catch|throw)\b', code)
    operands = re.findall(r'\b[a-zA-Z_][a-zA-Z0-9_]*\b', code)
    return operators, operands

def count_halstead_metrics(directory, extensions):
    operator_count = {}
    operand_count = {}
    
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith(extensions):
                file_path = os.path.join(root, file)
                try:
                    with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                        code = f.read()
                        operators, operands = extract_operators_and_operands(code)
                        
                        for op in operators:
                            operator_count[op] = operator_count.get(op, 0) + 1
                        
                        for opd in operands:
                            operand_count[opd] = operand_count.get(opd, 0) + 1
                except Exception as e:
                    print(f"Error reading {file_path}: {e}")
    
    # Calculate Halstead Metrics
    n1 = len(operator_count)  # Number of unique operators
    n2 = len(operand_count)   # Number of unique operands
    N1 = sum(operator_count.values())  # Total occurrences of operators
    N2 = sum(operand_count.values())   # Total occurrences of operands
    
    n = n1 + n2  # Vocabulary
    N = N1 + N2  # Length
    V = N * (n1 + n2) if n > 0 else 0  # Volume
    D = (n1 / 2) * (N2 / n2) if n2 > 0 else 0  # Difficulty
    E = D * V  # Effort
    B = V / 3000  # Estimated number of bugs
    
    return {
        "n1 (Unique Operators)": n1,
        "n2 (Unique Operands)": n2,
        "N1 (Total Operators)": N1,
        "N2 (Total Operands)": N2,
        "Vocabulary (n)": n,
        "Length (N)": N,
        "Volume (V)": V,
        "Difficulty (D)": D,
        "Effort (E)": E,
        "Estimated Bugs (B)": B
    }

if __name__ == "__main__":
    project_directory = input("Enter the project directory path: ")
    extensions = (".css", ".js", ".php", ".html")
    results = count_halstead_metrics(project_directory, extensions)
    
    print("\nHalstead Metrics Results:")
    for key, value in results.items():
        print(f"{key}: {value}")
